README.md
# ClipboardManager

* [VERSION 4.2.230117](https://github.com/carlosjhr64/clipboard_manager/releases)
* [github](https://github.com/carlosjhr64/clipboard_manager)
* [rubygems](https://rubygems.org/gems/clipboard_manager)

## DESCRIPTION:

Ruby Gtk3App Clipboard Manager.

## SCREENSHOT:

![screenshot](img/screenshot.png)

## FEATURES

* eval-calculator
* Wiktionary
* xdg-open url
* espeak

Also:

* History
* QR-Code copy to clipboard

## INSTALL:
```shell
$ gem install clipboard_manager
```
## MOUSE CLICKS ON LOGO:

+ Button #1: Toggle Minime
+ Button #2: Toggle On/Off
+ Button #3: Pop-up Application Menu

## CONFIGURATION:

After an initial run, your user configuration will found in:

    ~/.config/gtk3app/clipboardmanager/config-?.?.rbon

At top of the file you will find the available tasks:
```ruby
{
  Tasks!: {
    calculator: [
      "^([\\d\\.\\+\\-\\*\\/\\%\\(\\) ]{3,80})$",
      :reply,
      true
    ],
    dictionary: [
      "^(\\w+)$",
      :bashit,
      true,
      "xdg-open 'https://en.wiktionary.org/wiki/$1'"
    ],
    url: [
      "^https?://\\w[\\-\\+\\.\\w]*(\\.\\w+)(:\\d+)?(/\\S*)?$",
      :open,
      true
    ],
    espeak: [
      ".{40,}",
      :espeak,
      true
    ]
  },
  # ...
}
```
It is by this configuration that one can modify and add tasks.
Note that this is an [RBON](https://rubygems.org/gems/rbon) file.
ClipboardManager has four tasks methods:

* `:bashit` will take a command to be run by the system.
* `:open` will `xdg-open` the clip.
* `:espeak` will `espeak` the clip.
* `:reply` will `eval` the clip and display a message with the result.

With the boolean `true` value the clipboard will clear on the matched task.
If you don't want the clipboard cleared on a matched task,
set the boolean value to `false`.

Note that `:bashit` requires a extra command string which
it will substitute $0, $1, $2... with match data.
It then passes the string to system.

The `:caculator` task will run when the clip looks like a bunch of number being operated.

The `:espeak` task will run when the clip is at least 80 characters long.
It will have espeak read the text.

## HELP:
```console
$ clipboard_manager --help
Usage:
  clipboard_manager [:options+]
Options:
  -h --help
  -v --version
  --minime      	 Real minime
  --notoggle    	 Minime wont toggle decorated and keep above
  --notdecorated	 Dont decorate window
```
## LICENSE:

(The MIT License)

Copyright (c) 2023 CarlosJHR64

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 0.0.0 / 2014-11-27 10:03:08 -0800
=== 0.0.0 / 2014-11-29 14:04:29 -0800
9209ad9461eaa5f8d71ba576d681537b  pkg/clipboard_manager-0.0.0.gem
=== 1.0.0 / 2014-11-30 10:29:04 -0800
478f0a7f530a2c9fbc6d58cc4698ab3b  pkg/clipboard_manager-1.0.0.gem
=== 1.0.1 / 2014-11-30 10:45:43 -0800
7638ffbbee27c2d818897c16420f5aae  pkg/clipboard_manager-1.0.1.gem
=== 1.0.2 / 2014-12-13 12:37:29 -0800
a22c4fec8524bac407e0d6ab20d5fbb7  pkg/clipboard_manager-1.0.2.gem
=== 1.0.3 / 2014-12-18 15:24:34 -0800
7744da72d2c337151ed0f1ad42688a1d  pkg/clipboard_manager-1.0.3.gem
=== 1.0.4 / 2014-12-29 10:46:44 -0800
4047866390840bf0eec585517babdb93  pkg/clipboard_manager-1.0.4.gem
=== 1.0.5 / 2014-12-30 12:30:46 -0800
c05e4c832f101342535412409e66f436  pkg/clipboard_manager-1.0.5.gem
=== 1.1.0 / 2015-12-22 15:29:11 -0800
9122f1b3ccef489321ece83474208034  pkg/clipboard_manager-1.1.0.gem
=== 1.1.1 / 2015-12-29 11:16:52 -0800
dbf6b4929dfe084f788d9c4739d4dc57  pkg/clipboard_manager-1.1.1.gem
=== 2.0.0 / 2017-12-15 13:59:55 -0800
056d6fdd5fdad788b9ff3eea4fa2dc36  pkg/clipboard_manager-2.0.0.gem
=== 3.0.0 / 2018-12-19 06:23:59 -0800
dc3f4c734666d67ea5d1d2df8aae5197  pkg/clipboard_manager-3.0.0.gem
# 4.0.210706 / 2021-07-06 16:29:54 -0700
069ca1bb8d98dbd609e29960d6f9e201  pkg/clipboard_manager-4.0.210706.gem
# 4.1.230115 / 2023-01-15 13:05:20 -0800
593fea45acbe993998c7d2af1bc94d3f  pkg/clipboard_manager-4.1.230115.gem
# 4.2.230117 / 2023-01-17 06:31:56 -0800
8cb55e1afb1dc97ea14336c93704ad8e  pkg/clipboard_manager-4.2.230117.gem


lib/clipboard_manager/clipboard_manager.rb
def Gtk3App.toggle!
  @clipboard_manager_hook.do_toggle!
end

def Gtk3App.clipboard_manager_hook(hook)
  @clipboard_manager_hook = hook
end

class ClipboardManager
  using Rafini::Exception

  class NoYes < Such::Dialog
    def initialize(key)
      super
      add_button '_No', Gtk::ResponseType::CANCEL
      add_button '_Yes', Gtk::ResponseType::OK
    end

    def label(key)
      Such::Label.new child, key
    end

    def ok?
      show_all
      response = run
      destroy
      response == Gtk::ResponseType::OK
    end
  end

  class CancelOk < Such::Dialog
    def initialize(key)
      super
      add_button(Gtk::Stock::CANCEL, Gtk::ResponseType::CANCEL)
      add_button(Gtk::Stock::OK, Gtk::ResponseType::OK)
    end

    def combo(key)
      Such::ComboBoxText.new child, key
    end

    def runs
      show_all
      response = run
      yield if response == Gtk::ResponseType::OK
      destroy
    end
  end

  class Message < Such::Dialog
    def initialize(key)
      super
      add_button '_OK', Gtk::ResponseType::OK
    end

    def label(key)
      Such::Label.new child, key
    end

    def runs
      show_all
      response = run
      destroy
    end
  end

  CLIPBOARD = Gtk::Clipboard.get(Gdk::Selection::CLIPBOARD)

  def initialize(stage, toolbar, options)
    @image = toolbar.parent.children[0].child # Expander:hbox:EventImage:Image
    @timer = nil

    @working = GdkPixbuf::Pixbuf.new(file: CONFIG[:Working])
    @ok      = GdkPixbuf::Pixbuf.new(file: CONFIG[:Ok])
    @nope    = GdkPixbuf::Pixbuf.new(file: CONFIG[:Nope])
    @ready   = GdkPixbuf::Pixbuf.new(file: CONFIG[:Ready])
    @off     = GdkPixbuf::Pixbuf.new(file: CONFIG[:Off])

    @is_pwd  = Regexp.new(CONFIG[:IsPwd], Regexp::EXTENDED)

    vbox = Such::Box.new(stage, :vbox!)
    @running = Such::CheckButton.new(vbox, :running!, 'toggled'){toggled}
    @running.active = true

    @ask = Such::CheckButton.new vbox, :ask!
    @ask.active = true

    Such::Label.new vbox, :tasks!

    @checks = {}
    CONFIG[:Tasks!].keys.each do |key|
      @checks[key] = Such::CheckButton.new(vbox, [key.to_s.capitalize], {set_active: true})
    end

    Such::Button.new(vbox, :history_button!){do_history!}
    Such::Button.new(vbox, :qrcode_button!){do_qrcode!} if ClipboardManager.do_qrcode

    @history, @previous = [], nil
    text = request_text
    if text
      add_history text
      @previous = text
    end
    GLib::Timeout.add(CONFIG[:Sleep]) do
      step if @running.active?
      true # repeat
    end

    Gtk3App.clipboard_manager_hook(self)
    Gtk3App.logo_press_event do |button|
      case button
      when 1
        Gtk3App.minime!
      when 2
        Gtk3App.toggle!
      # When 3 gets captured by Gtk3App's main menu
      end
    end

    status(@ready)
  end

  # https://github.com/ruby-gnome2/ruby-gnome2/blob/master/gtk3/sample/misc/dialog.rb
  def do_history!
    dialog = CancelOk.new(:history_dialog!)
    Gtk3App.transient dialog
    combo = dialog.combo :history_combo!
    @history.each do |str|
      if str.length > CONFIG[:MaxString]
        n = CONFIG[:MaxString]/2 - 1
        str = "#{str[0..n]}...#{str[-n..-1]}"
      end
      combo.append_text(str)
    end
    dialog.runs do
      if combo.active_text
        @previous = nil
        CLIPBOARD.text = @history[combo.active]
      end
    end
  end

  def do_qrcode!
    qrcode = Timeout::timeout(CONFIG[:QrcTimeOut]){ Helpema::ZBar.cam() }
    CLIPBOARD.text = qrcode
    status(@ok)
  rescue
    $!.puts
    CLIPBOARD.clear
    status(@nope)
  end

  def question?(name)
    return true unless @ask.active?
    dialog = NoYes.new :question_dialog!
    Gtk3App.transient dialog
    dialog.label(:question_label!).text = "Run #{name}?"
    dialog.ok?
  end

  def toggled
    @running.active? ? status(@ready) : status(@off)
  end

  def do_toggle!
    text = request_text
    @previous = text
    @running.active = !@running.active?
    ClipboardManager.kill_espeak unless @running.active?
  end

  def request_text
    if text = CLIPBOARD.wait_for_text
      text.strip!
      return text unless text.empty? or @is_pwd.match?(text)
    end
    nil
  end

  def status(type)
    @image.set_pixbuf(type)
    @timer = Time.now
  end

  def step
    if @timer and Time.now - @timer > CONFIG[:StatusTimeOut]
      @timer = nil
      status @ready
    end
    text = request_text
    unless text.nil? or @previous == text
      @previous = text
      status @working
      GLib::Timeout.add(0) do
        manage(text)
        false # don't repeat
      end
    end
  end

  def add_history(text)
    @history.unshift text
    @history.uniq!
    @history.pop if @history.length > CONFIG[:MaxHistory]
  end

  def manage(text)
    add_history text
    CONFIG[:Tasks!].each do |name, _|
      next unless @checks[name].active?
      rgx, mth, clr, str = _
      rgx = Regexp.new(rgx, Regexp::EXTENDED | Regexp::MULTILINE)
      if md=rgx.match(text) and question?(name)
        CLIPBOARD.clear if clr
        begin
          case mth
          when :espeak
            espeak(text)
          when :open
            open(text)
          when :bashit
            bashit(md, str)
          when :reply
            reply(text)
          else
            raise "Method #{mth} not implemented."
          end
          status(@ok)
        rescue RuntimeError
          $!.puts
          status(@nope)
        end
        return
      end
    end
    status(@nope)
  end

  def ClipboardManager.kill_espeak
    espeak = CONFIG[:Espeak].split.first
    system "killall --quiet #{espeak}"
  end
  Gtk3App.finalize{ClipboardManager.kill_espeak}
  def espeak(text)
    Rafini.thread_bang! do
      ClipboardManager.kill_espeak
      IO.popen(CONFIG[:Espeak], 'w'){_1.puts text.strip}
    end
  end

  def open(text)
    Process.detach spawn CONFIG[:Open], text
  end

  def bashit(md, str)
    (md.length-1).downto(0) do |i|
      str = str.gsub(/\$#{i}/, md[i] || '')
    end
    $stderr.puts str
    Process.detach spawn str
  end

  def reply(text)
    dialog = Message.new(:reply_dialog!)
    Gtk3App.transient dialog
    begin
      dialog.label(:reply_label!).text = text
      dialog.label(:reply_label!).text = "#{eval text}"
    rescue
      dialog.label(:reply_label!).text = $!.message
    end
    dialog.runs
  end
end


lib/clipboard_manager/config.rb
class ClipboardManager
  using Rafini::String # String#semantic
  extend Rafini::Empty # a0 and h0

  is_pwd =
'\A
  (?!\w+:\/\/)          # not like url
  (?!\/[a-z]+\/[a-z])   # not like linux path
  (?![a-z]+\/[a-z]+\/)  # not like relative path
  (?=.*\d)              # at least one digit
  (?=.*[a-z])           # at least one lower case letter
  (?=.*[A-Z])           # at least one upper case letter
  (?=.*[^\w\s])         # at least one special character
  \S*$                  # no spaces
\Z'

  CONFIG = {
    Tasks!: { # Note that Ruby's Hash preserves order, and order here is important.
      calculator: [
        '^([\d\.\+\-\*\/\%\(\) ]{3,80})$',
        :reply,
        true, # clears clipboard
      ],
      dictionary: [
        '^(\w+)$',
        :bashit,
        true, # clears clipboard
        "xdg-open 'https://en.wiktionary.org/wiki/$1'",
      ],
      url: ['^https?://\w[\-\+\.\w]*(\.\w+)(:\d+)?(/\S*)?$', :open, true],
      espeak: ['.{40,}', :espeak, true],
    },

    StatusTimeOut: 3,
    Sleep: 750,
    MaxHistory: 13,
    MaxString: 60,

    QrcTimeOut: 12,

    IsPwd: is_pwd,

    # The text-to-speech needs to be able to receive text from stdin
    Espeak: 'espeak -s 180 --stdin',

    Working: "#{UserSpace::XDG['data']}/gtk3app/clipboardmanager/working.png",
    Ok:      "#{UserSpace::XDG['data']}/gtk3app/clipboardmanager/ok.png",
    Nope:    "#{UserSpace::XDG['data']}/gtk3app/clipboardmanager/nope.png",
    Ready:   "#{UserSpace::XDG['data']}/gtk3app/clipboardmanager/ready.png",
    Off:     "#{UserSpace::XDG['data']}/gtk3app/clipboardmanager/off.png",

    HelpFile: "https://github.com/carlosjhr64/clipboard_manager",
    Logo: "#{UserSpace::XDG['data']}/gtk3app/clipboardmanager/logo.png",

    ### Gui Things ###

    about_dialog: {
      set_program_name: 'Clipboard Manager',
      set_version: VERSION.semantic(0..1),
      set_copyright: '(c) 2023 CarlosJHR64',
      set_comments: 'A Ruby Gtk3App Clipboard Manager ',
      set_website: 'https://github.com/carlosjhr64/clipboard_manager',
      set_website_label: 'See it at GitHub!',
    },

    window: {
      set_title: "Clipboard Manager",
      set_window_position: :center,
    },

    VBOX: [:vertical],
    vbox: h0,
    vbox!: [:VBOX, :vbox],

    ASK: ['Ask For Confirmation'],
    ask: h0,
    ask!: [:ASK, :ask],

    RUNNING: ['On/Off'],
    running: h0,
    running!: [:RUNNING, :running],

    TASKS: ['Tasks:'],
    tasks: h0,
    tasks!: [:TASKS, :tasks],

    HISTORY_BUTTON: [label: 'History'],
    history_button: h0,
    history_button!: [:HISTORY_BUTTON, :history_button],

    HISTORY_DIALOG: a0,
    history_dialog: h0,
    history_dialog!: [:HISTORY_DIALOG, :history_dialog],

    HISTORY_COMBO: a0,
    history_combo: h0,
    history_combo!: [:HISTORY_COMBO, :history_combo],

    QRCODE_BUTTON: [label: 'QR-Code'],
    qrcode_button: h0,
    qrcode_button!: [:QRCODE_BUTTON, :qrcode_button],

    QUESTION_DIALOG: a0,
    question_dialog: {
      set_keep_above: true,
    },
    question_dialog!: [:question_dialog, :QUESTION_DIALOG],

    QUESTION_LABEL: a0,
    question_label: h0,
    question_label!: [:question_label, :QUESTION_LABEL],

    REPLY_LABEL: a0,
    reply_label: h0,
    reply_label!: [:reply_label, :REPLY_LABEL],

    REPLY_DIALOG: a0,
    reply_dialog: h0,
    reply_dialog!: [:reply_dialog, :REPLY_DIALOG],

    # Toggle's app-menu item.
    # Application MAY modify :TOGGLE for language.
    TOGGLE: [label: 'Toggle On/Off'],
    toggle: h0,
    toggle!: [:TOGGLE, :toggle, 'activate'],
    app_menu: {
      add_menu_item: [ :toggle!, :minime!, :help!, :about!, :quit!  ],
    },
  }
end
